chain/client/src/client.rs

```diff
@@ -455,6 +441,90 @@ impl Client {
         Ok(Some((encoded_chunk, merkle_paths, outgoing_receipts)))
     }
 
+    /// Pulls transactions from the transaction pool for a given shard up to the gas_limit.
+    /// Filters transactions one by one by persisting the state across valid validations.
+    fn get_filtered_transactions(
+        &mut self,
+        next_height: u64,
+        prev_block_timestamp: u64,
+        shard_id: u64,
+        chunk_extra: &ChunkExtra,
+        prev_block_header: &BlockHeader,
+    ) -> (usize, Vec<SignedTransaction>) {
+        // Total number of transactions pulled from the pool.
+        let mut num_checked_transactions = 0;
+        let mut transactions = vec![];
+        if let Some(mut iter) = self.shards_mgr.get_pool_draining_iterator(shard_id) {
+            // Total amount of gas burnt for converting transactions towards receipts.
+            let mut total_gas_burnt = 0;
+            // TODO: Update gas limit for transactions
+            let transactions_gas_limit = chunk_extra.gas_limit / 2;
+
+            let transaction_validity_period = self.chain.transaction_validity_period;
+            let mut state_update =
+                self.runtime_adapter.get_state_update(chunk_extra.state_root.clone());
+
+            // TODO: Check if we still need to use `block_expected_weight` from the config.
+            while transactions.len() < self.config.block_expected_weight as usize
+                && total_gas_burnt < transactions_gas_limit
+            {
+                // Whether we need to pull the transaction with the same key as before. It's true
+                // if the previous transaction was invalid.
+                let mut same_key = false;
+                let tx_with_gas_burnt = loop {
+                    if let Some(tx) = iter.next(same_key) {
+                        num_checked_transactions += 1;
+                        same_key = true;
+                        // Verifying the transaction is on the same chain and hasn't expired yet.
+                        if self
+                            .chain
+                            .mut_store()
+                            .check_blocks_on_same_chain(
+                                &prev_block_header,
+                                &tx.transaction.block_hash,
+                                transaction_validity_period,
+                            )
+                            .is_err()
+                        {
+                            continue;
+                        } else {
+                            // Verifying the validity of the transaction based on the current state.
+                            match self.runtime_adapter.validate_tx(
```

Comment:
this is really weird, that validate_tx requires a state_update

Reply:
Suggestions?

Reply:
Will revert validate back to the original state_root.